libafl/src/inputs/value.rs

```diff
@@ -0,0 +1,271 @@
+impl Input for ValueInput<Vec<u8>> {
+    fn generate_name(&self, _id: Option<CorpusId>) -> String {
+        format!("{:016x}", generic_hash_std(self))
+    }
+
+    /// Write this input to the file
+    #[cfg(feature = "std")]
+    fn to_file<P>(&self, path: P) -> Result<(), Error>
+    where
+        P: AsRef<Path>,
+    {
+        write_file_atomic(path, self.as_ref())?;
+        Ok(())
+    }
+
+    /// Load the content of this input from a file
+    #[cfg(feature = "std")]
+    fn from_file<P>(path: P) -> Result<Self, Error>
+    where
+        P: AsRef<Path>,
+    {
+        let mut file = File::open(path)?;
+        let mut data = vec![];
+        file.read_to_end(&mut data)?;
+        Ok(data.into())
+    }
+}
+
+impl<I> Numeric for ValueInput<I>
+where
+    I: Numeric,
+{
+    fn flip_all_bits(&mut self) {
+        self.as_mut().flip_all_bits();
+    }
+
+    fn flip_bit_at(&mut self, rhs: usize) {
+        self.as_mut().flip_bit_at(rhs);
+    }
+
+    fn wrapping_inc(&mut self) {
+        self.as_mut().wrapping_inc();
+    }
+
+    fn wrapping_dec(&mut self) {
+        self.as_mut().wrapping_dec();
+    }
+
+    fn twos_complement(&mut self) {
+        self.as_mut().twos_complement();
+    }
+
+    fn randomize<R: Rand>(&mut self, rand: &mut R) {
+        self.as_mut().randomize(rand);
+    }
+}
+
+/// Input type that holds a mutable reference to an inner value
+#[derive(Debug)]
+pub struct ValueMutRefInput<'a, I>(&'a mut I);
+
+impl<I> Deref for ValueMutRefInput<'_, I> {
+    type Target = I;
+
+    fn deref(&self) -> &Self::Target {
+        self.0
+    }
+}
+
+impl<I> DerefMut for ValueMutRefInput<'_, I> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.0
+    }
+}
+
+impl<'a, I> From<&'a mut I> for ValueMutRefInput<'a, I> {
+    fn from(value: &'a mut I) -> Self {
+        Self(value)
+    }
+}
+
+impl<'a, I> From<&'a mut ValueInput<I>> for ValueMutRefInput<'a, I> {
+    fn from(value: &'a mut ValueInput<I>) -> Self {
+        Self(value.as_mut())
+    }
+}
+
+impl<I> MappedInput for ValueMutRefInput<'_, I> {
+    type Type<'a>
+        = ValueMutRefInput<'a, I>
+    where
+        Self: 'a;
+}
+
+impl<I> Numeric for ValueMutRefInput<'_, I>
+where
+    I: Numeric,
+{
+    fn flip_all_bits(&mut self) {
+        self.deref_mut().flip_all_bits();
+    }
+
+    fn flip_bit_at(&mut self, rhs: usize) {
+        self.deref_mut().flip_bit_at(rhs);
+    }
+
+    fn wrapping_inc(&mut self) {
+        self.deref_mut().wrapping_inc();
+    }
+
+    fn wrapping_dec(&mut self) {
+        self.deref_mut().wrapping_dec();
+    }
+
+    fn twos_complement(&mut self) {
+        self.deref_mut().twos_complement();
+    }
+
+    fn randomize<R: Rand>(&mut self, rand: &mut R) {
+        self.deref_mut().randomize(rand);
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::{ValueInput, ValueMutRefInput};
+    use crate::mutators::numeric::Numeric;
+
+    fn take_numeric<I: Numeric + Clone>(i: I) {
```

Comment:
What is this supposed to test? I wouldn't be at all surprised if the compiler notices that this is all dead code and optimizes it away. Perhaps this should `assert_eq!` that `i.clone.f()` is the same as `ValueInput::from(i.clone()).f().inner()`?

Reply:
Valid point, this definitely is not a robust testing regime. 

To be honest, this was mostly a test during development that the types work out, so the fact that it compiles is the main testing success.

Previous implementations panicked on certain values, and especially min and max values seemed unstable. I don't think this is still possible, but I figured the cost of leaving this here is minimal compared to potential breakage.

Ideally however, we'd prove some of these properties, I have no idea how to do this though.

Reply:
I have renamed the test to better reflect what it's currently doing.

Reply:
I'm now also checking if the operations actually change the values. Still not comprehensive, but it's a start.